### PR TITLE
WIP: #54281 speedup pkg5 module

### LINKED ISSUE #54281: pkg5 module - Speed up installing lists of packages (Check for all packages at once)
##### SUMMARY
As of right now, pkg5 processes a list of packages to install by filtering the packages by checking for each package if it is installed.

I measured the time it takes checking ten packages doing it this way. On my system it was 5 seconds.
I then measured the time it takes checking once for all ten packages, which was below 1 second.

I propose to change the way the pkg5 module checks when installing packages. I'll add a pull request for this later.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
pkg5

##### ADDITIONAL INFORMATION
This would greatly speed up the installation of multiple packages (the more packages the faster it is)
```bash
# Check for each package in a separate command
root@dev:/export/home/vagrant# time $(pkg list fh;pkg list gr;pkg list sfs;pkg list vedg;pkg list segj;pkg list adgf;pkg list aeg;pkg list ws;pkg list wf)

pkg list: no packages matching the following patterns are installed:
  fh

pkg list: no packages matching the following patterns are installed:
  gr

pkg list: no packages matching the following patterns are installed:
  sfs

pkg list: no packages matching the following patterns are installed:
  vedg

pkg list: no packages matching the following patterns are installed:
  segj

pkg list: no packages matching the following patterns are installed:
  adgf

pkg list: no packages matching the following patterns are installed:
  aeg

pkg list: no packages matching the following patterns are installed:
  ws

pkg list: no packages matching the following patterns are installed:
  wf

real    0m5,259s
user    0m3,542s
sys     0m1,428s
```

```bash
# Check for all packages in one command
root@dev:/export/home/vagrant# time $(pkg list fh gr sfs vedg segj adgf aeg ws wf)

pkg list: no packages matching the following patterns are installed:
  adgf
  aeg
  fh
  gr
  segj
  sfs
  vedg
  wf
  ws

real    0m0,599s
user    0m0,404s
sys     0m0,166s
```

### PR DESCRIPTION
**Pull request by danowar2k**
##### SUMMARY
This resolves #54281.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
pkg5

##### ADDITIONAL INFORMATION
Added a function: Given a list of packages, it returns a dictionary of packages that are installed, packages that are not installed and the subset of the installed packages that could be updated.
Parts of the dictionary are then fed to the ensure function. I removed unnecessary helper functions.

This greatly improves speed of package checks, as not every package is checked alone but all are checked at once.

### CODE REVIEW

**Comment by danowar2k** [comment]
Wait, this isn't right yet....

**Comment by danowar2k** [comment]
Now this should work

**Comment by danowar2k** [comment]
Wait with reviewing, this seems to still have problems with FMRIs containing wildcards ("*"). I'll have to look into the FMRI syntax...

**Comment by danowar2k** [comment]
Additional question: Is there any testing done using Solaris 11 systems? Could I be of service setting that up, too? I don't yet understand your testing process, though...

**Comment by danowar2k** [comment]
Also, it seems the module had issues with installing FMRIs with wildcards before I started...

https://github.com/ansible/ansible/blob/0ec2d71b4a5ecd882b24bcd9498208b269d3a403/lib/ansible/modules/packaging/os/pkg5.py#L119-L122

https://github.com/ansible/ansible/blob/0ec2d71b4a5ecd882b24bcd9498208b269d3a403/lib/ansible/modules/packaging/os/pkg5.py#L171-L173

```bash
# If no pkg FMRIs containing "package_substring" are installed, this will return rc 1
# If any pkg FMRI containing "package_substring" is installed, this will return rc 0
# even if another pkg FMRI containing "package_substring" is not installed yet
pkg list -- *package_substring*
```
So installing packages using wildcards basically doesn't work yet. Let's see if I can fix that, too...That will get complicated though.

I'll try with pkg list -a and parsing both stdout and stderr and checking the requested FMRIs for wildcards.

**Comment by mator** [comment]
solaris 11 `pkg list` return codes:

* packages installed:
```
user@solaris11:~$ pkg list curl wget
NAME (PUBLISHER)                                  VERSION                    IFO
web/curl                                          7.60.0-11.4.0.0.1.14.0     i--
web/wget                                          1.19.5-11.4.0.0.1.14.0     i--
user@solaris11:~$ echo $?
0
```
* packages not installed:
```
user@solaris11:~$ pkg list package1 package2

pkg list: no packages matching the following patterns are installed:
  package1
  package2
user@solaris11:~$ echo $?
1
```
* some packages installed and some is not:
```
user@solaris11:~$ pkg list curl package1 package2 wget
NAME (PUBLISHER)                                  VERSION                    IFO
web/curl                                          7.60.0-11.4.0.0.1.14.0     i--
web/wget                                          1.19.5-11.4.0.0.1.14.0     i--

pkg list: no packages matching the following patterns are installed:
  package1
  package2
user@solaris11:~$ echo $?
3
```

File: lib/ansible/modules/packaging/os/pkg5.py [hunk 1]
```diff
@@ -101,36 +101,25 @@ def main():
...
-            'subcommand': 'install',
-        },
-        'absent': {
-            'filter': lambda p: is_installed(module, p),
-            'subcommand': 'uninstall',
-        },
-    }
+    subcommand = ['install']
+    if state == 'absent':
+        subcommand = ['uninstall']
```
Comment (mator) [inline lib/ansible/modules/packaging/os/pkg5.py]: behaviour change without documenting it

**Review by mator — CHANGES_REQUESTED** [review]
:deleted

**Comment by danowar2k** [comment]
Sorry, I'm still changing code but forgot to close the PR in the meantime...
Once I'm done, I'm gonna reopen this one, document Solaris pkg behaviour, the old behaviour of the pkg5 module and the new behaviour of the pkg5 module here.

**Comment by dagwieers** [comment]
@danowar2k Before we can accept any changes to existing modules, we need to have integration tests so it gives some confidence this is not breaking existing functionality. Especially since not a lot of people in the community are involved in the Ansible community.

See: https://github.com/ansible/community/wiki/Solaris

**Comment by danowar2k** [comment]
Okay, the commit message contains a description of the changes.

Now to document pkg behaviour:
pkg install FMRI_PATTERN

Variant A:
- pattern contains no version (no "@" part)
- pattern contains @latest
-> **requested_package_version**=latest

Variant B:
- pattern contains "@specific_version_part"
-> **requested_package_version**=specific_version_part

- FMRI_PATTERN without wildcards
  - no matching package
    - stderr contains "  FMRI_PATTERN\n"
  - one matching package, not installed
     - Variant A: installs **requested_package_version** version of package if allowed
     - Variant B: installs **requested_package_version** if allowed, fail otherwise (can't update to version)
  - one matching package, installed but not the **requested_package_version**
     - Variant A: updates to the **requested_package_version**
     - Variant B: updates/downgrades to the **requested_package_version** if allowed, fail otherwise
  - one matching package, installed in the **requested_package_version**
     - "Nothing to do"
  - multiple matching packages
     - fails with error: ambiguous because multiple packages match

- FMRI_PATTERN with wildcards
  - no matching package
    - stderr contains "  FMRI_PATTERN\n"
  - one matching package, not installed
     - installs **requested_package_version** of package if allowed
  - one matching package, installed but not **requested_package_version**
     - updates/downgrades to the **requested_package_version** if possible
  - one matching package, installed in **requested_package_version**
     - "Nothing to do"

  - multiple matching packages
     - for each matched package, proceed like "one matching package"

- multiple FMRI_PATTERNS matching the same package but different **requested_package_version**
  - fails with error: can't satisfy conditions

**Comment by danowar2k** [comment]
Okay, now to document the way the ansible pkg5 module used to work. I'll only document where the expected behaviour differed from the actual behaviour.

- FMRI_PATTERN without wildcards
  - one matching package, installed but not the **requested_package_version**
     - if `@latest` is specifically requested (even when state is set to "present", which is also the default state, this is entirely possible to happen), nothing is done because is_installed returns true
  - multiple matching packages, none is installed
     - fails like usual
  - multiple matching packages, at least one is installed
     - is_installed returns true, no error is reported, user doesn't know multiple packages match

- FMRI_PATTERN with wildcards
  - one matching package, installed but not **requested_package_version**
     - if `@latest` is specifically requested (even when state is set to "present", which is also the default state, this is entirely possible to happen), nothing is done because is_installed returns true
  - multiple matching packages, none is installed
     - installs all matching packages in the requested version if possible, fails otherwise (installs all or nothing)
  - multiple matching packages, at least one is installed
     - is_installed returns true, no error is reported, user doesn't know multiple packages match and the not installed package is never installed

**Comment by danowar2k** [comment]
A problem seems to present itself: My pull request was for speeding up the pkg5 module, but some other bugs seem to have come up. What do I do when that happens and I think my code does both (fixes bugs and speeds up pkg5)?

**Comment by danowar2k** [comment]
Example for bugs:
Use pkg5 module in Ansible playbook on a Solaris VM:

```
# This won't update the system because entire is always installed in some version
- hosts: all
  name: Run the pkg5 module for tests
  tasks:
    - name: Explicitly install the latest entire package (Do a Solaris upgrade)
      become: true
      pkg5:
        name:
          - entire@latest
        state: present

# This won't update the system because entire is always installed in some version
    - name: Install the latest entire package without mentioning the state
      become: true
      pkg5:
        name:
          - entire@latest

# This won't update the system because entire is always installed in some version
    - name: Install the latest entire package with wildcards
      pkg5:
        name:
          - "*entire*@latest"
        state: present

# Precondition:
# $ pkg list -a *dns/bind*
# NAME (PUBLISHER)                                  VERSION                    IFO
# network/dns/bind                                  9.10.8.1.0-11.4.3.0.1.1.0  i--
# service/network/dns/bind                          9.10.8.1.0-11.4.3.0.1.1.0  ---
# This won't fail with the below message, but instead will say "Nothing to do"
    - name: Install the one and only dns/bind package (but multiple packages match)
      pkg5:
        name:
          - "dns/bind"
        state: present

# This won't install anything because at least one matching package is installed
    - name: Install all packages matching dns/bind (won't install anything)
      pkg5:
        name:
          - "*dns/bind*"
        state: present
```
Example for ambiguous package name error:
```
# pkg install -n dns/bind

pkg install: 'dns/bind' matches multiple packages
        pkg://solaris/service/network/dns/bind
        pkg://solaris/network/dns/bind

Please provide one of the package FMRIs listed above to the install command.
```

**Comment by danowar2k** [comment]
How my pkg5 module works different:

- State "latest" just tries to install/update all matching packages to latest
   - Discussion: Should the combination of state:"latest" + FMRI patterns with specific requested versions lead to an error message ("Why would you do that? Conflicting intentions?")
- State "present" installs all matching not installed packages to the requested versions, updates all packages matching versioned patterns to the requested versions if possible and leaves other installed packages alone

**Comment by danowar2k** [comment]
Okay, I'm done for now. Please tell me if and where I can assist.

**Comment by danowar2k** [comment]
> @danowar2k Before we can accept any changes to existing modules, we need to have integration tests so it gives some confidence this is not breaking existing functionality. Especially since not a lot of people in the community are involved in the Ansible community.
> 
> See: https://github.com/ansible/community/wiki/Solaris

I've read up on testing Ansible in the docs. The integration testing page seems a bit more sparse than the other ones.

I'm trying to get a feeling for this. If I'm really trying to show that the module works as intended I think I need a generated Solaris VM to test on. I don't yet understand what I have to do to integrate a Solaris VM in your testing suite.

Meanwhile, I surely can restructure more of my code into OS-independent functions, leaving only the functions containing the run_command calls untested. I can mock output generated by pkg list etc. and then feed the functions this output. Would that be good for a start?

**Comment by danowar2k** [comment]
If I understand this right, you have a Shippable CI configuration which runs unit tests, sanity tests, etc. . It also declares tests on certain systems which seem to be using Docker containers. Solaris and Docker doesn't seem to work together, so no Docker image for Solaris exists. So I can't add that.

I'm developing playbooks for Solaris x86 machines using a pipeline of
Packer generated VirtualBox ISO + Vagrant box -> running the Vagrant box in VirtualBox -> applying Ansible playbooks to the VM / developing applications in the VM

If there is a CI provider that can run VirtualBox images and Vagrant, one could generate a second configuration for that CI provider to run Solaris specific tests on the started Solaris VirtualBox VM.

What do you think?

**Comment by danowar2k** [comment]
Gitlab can use VirtualBox:
https://docs.gitlab.com/runner/executors/virtualbox.html

Gitlab can be activated with Github:
https://docs.gitlab.com/ee/ci/ci_cd_for_external_repos/github_integration.html

Maybe a way to facilitate testing with Solaris VMs?

**Comment by mavit** [comment]
Have a look at https://github.com/ansible/ansible/pull/51651, which introduces tests for the `pkgutil` module.  In this example a pre-existing Solaris host is required to test with.  This is no good for automated testing (since we don't have any Solaris VMs dedicated to CI (yet?)), but does mean that a human who has access to Solaris machines can run the tests manually before merging.

**Pull request closed without merging**